File: scripts/notion/clean-other-contacts.py
#!/usr/bin/env python3
import os
import json
import requests
import re
import time
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

NOTION_TOKEN = os.getenv('NOTION_TOKEN')
THERAPISTS_DB_ID = os.getenv('THERAPISTS_DATABASE_ID')

if not NOTION_TOKEN or not THERAPISTS_DB_ID:
    print("❌ Error: NOTION_TOKEN and THERAPISTS_DATABASE_ID required in .env")
    exit(1)

NOTION_VERSION = "2022-06-28"
HEADERS = {
    "Authorization": f"Bearer {NOTION_TOKEN}",
    "Notion-Version": NOTION_VERSION,
    "Content-Type": "application/json"
}

def get_property_value(prop):
    """Extract value from a Notion property"""
    if not prop:
        return ""
    
    prop_type = prop.get('type')
    
    if prop_type == 'title':
        return prop.get('title', [{}])[0].get('plain_text', '') if prop.get('title') else ''
    elif prop_type == 'rich_text':
        return prop.get('rich_text', [{}])[0].get('plain_text', '') if prop.get('rich_text') else ''
    elif prop_type == 'phone_number':
        return prop.get('phone_number', '')
    elif prop_type == 'url':
        return prop.get('url', '')
    elif prop_type == 'email':
        return prop.get('email', '')
    
    return ""

def parse_other_contacts(text, existing_phone):
    """Parse the Other Contacts field intelligently"""
    result = {
        'phone': None,
        'facebook': None,
        'instagram': None,
        'twitter': None,
        'linkedin': None,
        'should_clear': False
    }
    
    if not text or not text.strip():
        return result
    
    # Extract phone numbers (Australian format)
    phone_patterns = [
        r'(?:\+61|0)\s?[2-478]\s?\d{4}\s?\d{4}',
        r'\b0[2-478]\d{8}\b',
        r'\b04\d{2}\s?\d{3}\s?\d{3}\b'
    ]
    
    for pattern in phone_patterns:
        match = re.search(pattern, text)
        if match:
            found_phone = match.group(0)
            normalized_found = found_phone.replace(' ', '')
            normalized_existing = existing_phone.replace(' ', '') if existing_phone else ''
            
            if not normalized_existing or normalized_found != normalized_existing:
                result['phone'] = found_phone
                break
    
    # Parse social media
    # Combined pattern: "facebook & instagram is [name]"
    combined_match = re.search(r'(?:facebook|fb)\s*(?:&|and)\s*(?:instagram|ig|insta)\s*(?:is|:|-)\s*([a-z0-9_.]+)', text, re.IGNORECASE)
    
    if combined_match:
        username = combined_match.group(1)
        result['facebook'] = f"https://facebook.com/{username}"
        result['instagram'] = f"https://instagram.com/{username}"
    else:
        # Individual patterns
        fb_match = re.search(r'(?:facebook|fb)\s*(?:is|:|-)\s*([a-z0-9_.]+)', text, re.IGNORECASE)
        if fb_match:
            result['facebook'] = f"https://facebook.com/{fb_match.group(1)}"
        
        ig_match = re.search(r'(?:instagram|ig|insta)\s*(?:is|:|-)\s*([a-z0-9_.]+)', text, re.IGNORECASE)
        if ig_match:
            result['instagram'] = f"https://instagram.com/{ig_match.group(1)}"
    
    # Twitter
    twitter_match = re.search(r'(?:twitter|x)\s*(?:is|:|-)\s*@?([a-z0-9_]+)', text, re.IGNORECASE)
    if twitter_match:
        result['twitter'] = f"https://twitter.com/{twitter_match.group(1)}"
    
    # LinkedIn
    linkedin_match = re.search(r'linkedin\s*(?:is|:|-)\s*([a-z0-9\-]+)', text, re.IGNORECASE)
    if linkedin_match:
        result['linkedin'] = f"https://linkedin.com/in/{linkedin_match.group(1)}"
    
    # Determine if we should clear the field
    if any([result['phone'], result['facebook'], result['instagram'], result['twitter'], result['linkedin']]):
        result['should_clear'] = True
    elif existing_phone and existing_phone.lower().replace(' ', '') in text.lower().replace(' ', ''):
        result['should_clear'] = True  # It's just a duplicate phone number
    
    return result

def main():
    print("═" * 80)
    print("  Clean \"Other Contacts\" Column")
    print("  Applying changes to Notion database")
    print("═" * 80)
    print()
    print("🔍 Reading Victorian Therapists database...\n")
    
    # Query the database
    url = f"https://api.notion.com/v1/databases/{THERAPISTS_DB_ID}/query"
    all_results = []
    has_more = True
    start_cursor = None
    
    while has_more:
        payload = {"page_size": 100}
        if start_cursor:
            payload["start_cursor"] = start_cursor
        
        response = requests.post(url, headers=HEADERS, json=payload)
        
        if response.status_code != 200:
            print(f"❌ Error: {response.json()}")
            exit(1)
        
        data = response.json()
        all_results.extend(data.get('results', []))
        has_more = data.get('has_more', False)
        start_cursor = data.get('next_cursor')
    
    print(f"✅ Found {len(all_results)} total entries\n")
    print("═" * 80)
    print()
    
    processed_count = 0
    updated_count = 0
    skipped_count = 0
    
    for entry in all_results:
        props = entry.get('properties', {})
        page_id = entry.get('id')
        
        # Extract current values
        first_name = get_property_value(props.get('First Name'))
        last_name = get_property_value(props.get('Last Name'))
        fullname = get_property_value(props.get('Fullname'))
        
        # Construct the display name
        if first_name and last_name:
            name = f"{first_name} {last_name}"
        elif fullname:
            name = fullname
        elif first_name:
            name = first_name
        else:
            name = "Unknown"
        
        other_contacts = get_property_value(props.get('Other contact details, social media, etc.'))
        existing_phone = get_property_value(props.get('Phone'))
        existing_facebook = get_property_value(props.get('Facebook'))
        existing_instagram = get_property_value(props.get('Instagram'))
        existing_twitter = get_property_value(props.get('Twitter/X'))
        existing_linkedin = get_property_value(props.get('LinkedIn'))
        
        processed_count += 1
        
        if not other_contacts or not other_contacts.strip():
            continue
        
        # Parse the other contacts field
        parsed = parse_other_contacts(other_contacts, existing_phone)
        
        # Build update payload
        updates = {}
        has_updates = False
        
        # Only update if we found new data
        if parsed['phone'] and not existing_phone:
            updates['Phone'] = {'rich_text': [{'text': {'content': parsed['phone']}}]}
            has_updates = True
        
        if parsed['facebook'] and not existing_facebook:
            updates['Facebook'] = {'url': parsed['facebook']}
            has_updates = True
        
        if parsed['instagram'] and not existing_instagram:
            updates['Instagram'] = {'rich_text': [{'text': {'content': parsed['instagram']}}]}
            has_updates = True
        
        if parsed['twitter'] and not existing_twitter:
            updates['Twitter/X'] = {'rich_text': [{'text': {'content': parsed['twitter']}}]}
            has_updates = True
        
        if parsed['linkedin'] and not existing_linkedin:
            updates['LinkedIn'] = {'url': parsed['linkedin']}
            has_updates = True
        
        # Clear the "Other Contacts" field if we extracted data or it's a duplicate
        if parsed['should_clear']:
            updates['Other contact details, social media, etc.'] = {'rich_text': []}
            has_updates = True
        
        if has_updates:
            print(f"📝 Updating: {name}")
            print(f"   Original: \"{other_contacts}\"")
            
            if parsed['phone']:
                print(f"   ✓ Phone: {parsed['phone']}")
            if parsed['facebook']:
                print(f"   ✓ Facebook: {parsed['facebook']}")
            if parsed['instagram']:
                print(f"   ✓ Instagram: {parsed['instagram']}")
            if parsed['twitter']:
                print(f"   ✓ Twitter: {parsed['twitter']}")
            if parsed['linkedin']:
                print(f"   ✓ LinkedIn: {parsed['linkedin']}")
            if parsed['should_clear']:
                print(f"   ✓ Clearing \"Other Contacts\" field")
            
            # Update the page
            update_url = f"https://api.notion.com/v1/pages/{page_id}"
            update_payload = {"properties": updates}
            
            update_response = requests.patch(update_url, headers=HEADERS, json=update_payload)
            
            if update_response.status_code == 200:
                print(f"   ✅ Updated successfully")
                updated_count += 1
            else:
                print(f"   ❌ Error: {update_response.json()}")
            
            print()
            
            # Rate limiting - be nice to Notion API (~3 requests/second)
            time.sleep(0.334)
        else:
            skipped_count += 1
    
    print()
    print("═" * 80)
    print()
    print("📊 Summary:")
    print(f"   Total entries processed: {processed_count}")
    print(f"   Entries updated: {updated_count}")
    print(f"   Entries skipped: {skipped_count}")
    print()
    print("✅ Done!")

if __name__ == "__main__":
    main()

File: scripts/notion/clean-other-contacts.js
import { Client } from '@notionhq/client';
import * as dotenv from 'dotenv';

// Load environment variables
dotenv.config();

const NOTION_TOKEN = process.env.NOTION_TOKEN || process.env.NOTION_API_KEY;
const THERAPISTS_DB_ID = process.env.THERAPISTS_DATABASE_ID;

if (!NOTION_TOKEN || !THERAPISTS_DB_ID) {
    console.error('❌ Error: NOTION_TOKEN and THERAPISTS_DATABASE_ID required');
    process.exit(1);
}

const notion = new Client({ auth: NOTION_TOKEN });

// Helper function to extract property value
function extractPropertyValue(property) {
    if (!property) return '';
    try {
        switch (property.type) {
            case 'title':
                return property.title?.[0]?.plain_text || '';
            case 'rich_text':
                return property.rich_text?.[0]?.plain_text || '';
            case 'email':
                return property.email || '';
            case 'phone_number':
                return property.phone_number || '';
            case 'url':
                return property.url || '';
            default:
                return '';
        }
    } catch {
        return '';
    }
}

// Parse "Other Contacts" field intelligently
function parseOtherContacts(otherContactsText, existingPhone) {
    const result = {
        phone: null,
        facebook: null,
        instagram: null,
        twitter: null,
        linkedin: null,
        shouldClear: false
    };

    if (!otherContactsText || otherContactsText.trim() === '') {
        return result;
    }

    const text = otherContactsText.toLowerCase().trim();
    
    // Extract phone numbers (Australian format)
    // Patterns: 0412 345 678, 0412345678, 04 1234 5678, +61 412 345 678
    const phonePatterns = [
        /(?:\+61|0)\s?[2-478]\s?\d{4}\s?\d{4}/g,
        /\b0[2-478]\d{8}\b/g,
        /\b04\d{2}\s?\d{3}\s?\d{3}\b/g
    ];
    
    for (const pattern of phonePatterns) {
        const matches = otherContactsText.match(pattern);
        if (matches && matches.length > 0) {
            // If there's already a phone number in the Phone column, skip duplicates
            const foundPhone = matches[0].replace(/\s/g, '');
            const normalizedExisting = existingPhone ? existingPhone.replace(/\s/g, '') : '';
            
            if (!normalizedExisting || foundPhone !== normalizedExisting) {
                result.phone = matches[0];
            }
        }
    }
    
    // Parse social media hints
    // "facebook & instagram is [name]" or "facebook: [name]"
    const fbPatterns = [
        /facebook\s*(?:&|and)?\s*instagram\s*(?:is|:|\-)\s*([a-z0-9_.]+)/i,
        /facebook\s*(?:is|:|\-)\s*([a-z0-9_.]+)/i,
        /fb\s*(?:is|:|\-)\s*([a-z0-9_.]+)/i
    ];
    
    const igPatterns = [
        /instagram\s*(?:&|and)?\s*facebook\s*(?:is|:|\-)\s*([a-z0-9_.]+)/i,
        /instagram\s*(?:is|:|\-)\s*([a-z0-9_.]+)/i,
        /ig\s*(?:is|:|\-)\s*([a-z0-9_.]+)/i,
        /insta\s*(?:is|:|\-)\s*([a-z0-9_.]+)/i
    ];
    
    // Check for "facebook & instagram is [name]" pattern first
    const combinedPattern = /(?:facebook|fb)\s*(?:&|and)\s*(?:instagram|ig|insta)\s*(?:is|:|\-)\s*([a-z0-9_.]+)/i;
    const combinedMatch = otherContactsText.match(combinedPattern);
    
    if (combinedMatch) {
        const username = combinedMatch[1];
        result.facebook = `https://facebook.com/${username}`;
        result.instagram = `https://instagram.com/${username}`;
    } else {
        // Check individual patterns
        for (const pattern of fbPatterns) {
            const match = otherContactsText.match(pattern);
            if (match) {
                result.facebook = `https://facebook.com/${match[1]}`;
                break;
            }
        }
        
        for (const pattern of igPatterns) {
            const match = otherContactsText.match(pattern);
            if (match) {
                result.instagram = `https://instagram.com/${match[1]}`;
                break;
            }
        }
    }
    
    // Twitter/X patterns
    const twitterPatterns = [
        /(?:twitter|x)\s*(?:is|:|\-)\s*@?([a-z0-9_]+)/i
    ];
    
    for (const pattern of twitterPatterns) {
        const match = otherContactsText.match(pattern);
        if (match) {
            result.twitter = `https://twitter.com/${match[1]}`;
            break;
        }
    }
    
    // LinkedIn patterns
    const linkedinPatterns = [
        /linkedin\s*(?:is|:|\-)\s*([a-z0-9\-]+)/i,
        /linkedin\.com\/in\/([a-z0-9\-]+)/i
    ];
    
    for (const pattern of linkedinPatterns) {
        const match = otherContactsText.match(pattern);
        if (match) {
            result.linkedin = `https://linkedin.com/in/${match[1]}`;
            break;
        }
    }
    
    // Determine if we should clear the field
    // Clear if we extracted meaningful data OR if it only contains duplicated phone
    if (result.phone || result.facebook || result.instagram || result.twitter || result.linkedin) {
        result.shouldClear = true;
    } else if (existingPhone && text.includes(existingPhone.toLowerCase().replace(/\s/g, ''))) {
        result.shouldClear = true; // It's just a duplicate phone number
    }
    
    return result;
}

async function cleanOtherContactsColumn() {
    try {
        console.log('🔍 Reading Victorian Therapists database...\n');
        
        // Get all entries
        let allResults = [];
        let hasMore = true;
        let startCursor = undefined;
        
        while (hasMore) {
            const response = await notion.databases.query({
                database_id: THERAPISTS_DB_ID,
                start_cursor: startCursor,
                page_size: 100
            });
            
            allResults = allResults.concat(response.results);
            hasMore = response.has_more;
            startCursor = response.next_cursor;
        }
        
        console.log(`✅ Found ${allResults.length} total entries\n`);
        console.log('═'.repeat(80));
        
        let processedCount = 0;
        let updatedCount = 0;
        let skippedCount = 0;
        
        for (const page of allResults) {
            processedCount++;
            const props = page.properties;
            
            // Get current values
            const name = extractPropertyValue(props['Fullname'] || props['Name']);
            const otherContacts = extractPropertyValue(props['Other contacts (specify if it is mobile, phone, facebook page, business name, etc)']);
            const existingPhone = extractPropertyValue(props['Phone']);
            const existingFacebook = extractPropertyValue(props['Facebook']);
            const existingInstagram = extractPropertyValue(props['Instagram']);
            
            if (!otherContacts || otherContacts.trim() === '') {
                skippedCount++;
                continue;
            }
            
            // Parse the other contacts field
            const parsed = parseOtherContacts(otherContacts, existingPhone);
            
            // Build update payload
            const updates = {};
            let hasUpdates = false;
            
            // Only update if we found new data
            if (parsed.phone && !existingPhone) {
                updates['Phone'] = { phone_number: parsed.phone };
                hasUpdates = true;
            }
            
            if (parsed.facebook && !existingFacebook) {
                updates['Facebook'] = { url: parsed.facebook };
                hasUpdates = true;
            }
            
            if (parsed.instagram && !existingInstagram) {
                updates['Instagram'] = { url: parsed.instagram };
                hasUpdates = true;
            }
            
            if (parsed.twitter && (!props['Twitter/X'] || !extractPropertyValue(props['Twitter/X']))) {
                updates['Twitter/X'] = { url: parsed.twitter };
                hasUpdates = true;
            }
            
            if (parsed.linkedin && (!props['LinkedIn'] || !extractPropertyValue(props['LinkedIn']))) {
                updates['LinkedIn'] = { url: parsed.linkedin };
                hasUpdates = true;
            }
            
            // Clear the "Other Contacts" field if we extracted data or it's a duplicate
            if (parsed.shouldClear) {
                updates['Other contacts (specify if it is mobile, phone, facebook page, business name, etc)'] = { rich_text: [] };
                hasUpdates = true;
            }
            
            if (hasUpdates) {
                console.log(`\n📝 Updating: ${name || 'Unknown'}`);
                console.log(`   Original: "${otherContacts}"`);
                
                if (parsed.phone) console.log(`   ✓ Phone: ${parsed.phone}`);
                if (parsed.facebook) console.log(`   ✓ Facebook: ${parsed.facebook}`);
                if (parsed.instagram) console.log(`   ✓ Instagram: ${parsed.instagram}`);
                if (parsed.twitter) console.log(`   ✓ Twitter: ${parsed.twitter}`);
                if (parsed.linkedin) console.log(`   ✓ LinkedIn: ${parsed.linkedin}`);
                if (parsed.shouldClear) console.log(`   ✓ Clearing "Other Contacts" field`);
                
                // Update the page
                await notion.pages.update({
                    page_id: page.id,
                    properties: updates
                });
                
                updatedCount++;
                console.log(`   ✅ Updated successfully`);
                
                // Rate limiting - be nice to Notion API
                await new Promise(resolve => setTimeout(resolve, 334)); // ~3 requests/second
            } else {
                console.log(`\n⏭️  Skipping: ${name || 'Unknown'} - no updates needed`);
                console.log(`   Content: "${otherContacts}"`);
                skippedCount++;
            }
        }
        
        console.log('\n' + '═'.repeat(80));
        console.log('\n📊 Summary:');
        console.log(`   Total entries processed: ${processedCount}`);
        console.log(`   Entries updated: ${updatedCount}`);
        console.log(`   Entries skipped: ${skippedCount}`);
        console.log('\n✅ Done!');
        
    } catch (error) {
        console.error('\n❌ Error:', error.message);
        if (error.body) {
            console.error('Details:', JSON.stringify(error.body, null, 2));
        }
        throw error;
    }
}

// Run the script
console.log('═'.repeat(80));
console.log('  Clean "Other Contacts" Column');
console.log('  Parse phone numbers and social media into proper columns');
console.log('═'.repeat(80));
console.log();

cleanOtherContactsColumn()
    .then(() => {
        console.log('\n🎉 All done!');
    })
    .catch((error) => {
        console.error('\n❌ Failed:', error.message);
        process.exit(1);
    });

